File: 2015/2015Robot/src/Commands/TurnSpecifiedDegreesCommand.cpp
#include "TurnSpecifiedDegreesCommand.h"
#include "math.h"

// If we're within this # of degrees to target, then we're good.

#define DEGREES_PRECISION 1.5


/*
 * This command turns the number of degrees specified in the dashboard field.
 */
TurnSpecifiedDegreesCommand::TurnSpecifiedDegreesCommand(float degreesToTurn) 
	: CommandBase("TurnSpecifiedDegreesCommand") {
	Requires(chassis);
	this->degreesToTurn = degreesToTurn;
}

// Called just before this Command runs the first time
void TurnSpecifiedDegreesCommand::Initialize() {
	this->finished = false;
	sensorSubsystem->GetIMU()->ZeroYaw();
	this->startingAngle = sensorSubsystem->GetIMU()->GetYaw();
	this->goalAngle = startingAngle + this->degreesToTurn;
	printf("TurnSpecifiedDegreesCommand %f\n", this->degreesToTurn);
}

// Called repeatedly when this Command is scheduled to run
void TurnSpecifiedDegreesCommand::Execute() {
	float currentAngle = sensorSubsystem->GetIMU()->GetYaw();
	float angleDifference = goalAngle - currentAngle;
	
	SmartDashboard::PutNumber("Goal angle", goalAngle);
	SmartDashboard::PutNumber("Angle difference", fabs(angleDifference));
	printf("Angle difference %f goalAngle %f currentAngle %f\n", angleDifference, goalAngle, currentAngle);
	
	float turnRate = 0;
	
	if (fabs(angleDifference) < DEGREES_PRECISION) {
		chassis->Stop();
		finished = true;
	} else {
		// We slow our rate of turn as we get close to the angle we want.
		// These values are guesses.  A PID would be better here.

		if(fabs(angleDifference) > 60.0)  {
			turnRate = 0.75;
		} else if(fabs(angleDifference) > 45.0) {
			turnRate = 0.45;
		} else if(fabs(angleDifference) > 20.0) {
			turnRate = 0.43;
		} else if(fabs(angleDifference) > 15.0) {
			turnRate = 0.40;
		} else {
			turnRate = 0.35;
		}

		//if (angleDifference > 10 || angleDifference < -10) {
		//	turnRate = 0.5;
		//} else {
		//	// Look at changing this at competition
		//	turnRate = 0.25;
		//}
		
		SmartDashboard::PutNumber("turn rate", turnRate);
		
		if (angleDifference > 0) {
			chassis->GetDrive()->ArcadeDrive(0.0, -turnRate);
		} else {
			chassis->GetDrive()->ArcadeDrive(0.0, turnRate);
		}
	}
}

// Make this return true when this Command no longer needs to run execute()
bool TurnSpecifiedDegreesCommand::IsFinished() {
	return finished;	
}

// Called once after isFinished returns true
void TurnSpecifiedDegreesCommand::End() {
	printf("TurnSpecifiedDegreesCommand completed.\n");
	chassis->Stop();
}

// Called when another command which requires one or more of the same
// subsystems is scheduled to run
void TurnSpecifiedDegreesCommand::Interrupted() {
	chassis->Stop();
}

File: 2016/DriveBot/src/com/kennedyrobotics/robot2016/commands/TurnSpecifiedDegreesCommand.java
package com.kennedyrobotics.robot2016.commands;

import com.kennedyrobotics.robot2016.Robot;

import edu.wpi.first.wpilibj.command.Command;

/**
 *
 */
public class TurnSpecifiedDegreesCommand extends Command {

	private static final double DEGREES_PRECISION = 2;
	
	private boolean finished = false;
	private double startingAngle = 0.0;
	private double goalAngle = 0.0;
	private final double degreesToTurn;
	
    public TurnSpecifiedDegreesCommand(double degreesToTurn) {
    	requires(Robot.chassis);
    	this.degreesToTurn = degreesToTurn;
    }

    // Called just before this Command runs the first time
    protected void initialize() {
    	finished = false;
    	startingAngle = Robot.sensorSubsystem.getYaw();
    	goalAngle = startingAngle + degreesToTurn;
    	
    	System.out.println("[TurnSpecifiedDegreesCommand] Going to turn " + degreesToTurn);
    }

    // Called repeatedly when this Command is scheduled to run
    protected void execute() {
    	
    	double currentAngle = Robot.sensorSubsystem.getYaw();
    	double angleDifference = goalAngle - currentAngle;
    	
    	System.out.println("Angle Difference " + angleDifference + " gaolAngle " + goalAngle + " currentAngle " + currentAngle);
    	
    	double turnRate = 0;
    	if(Math.abs(angleDifference) <  DEGREES_PRECISION) {
    		Robot.chassis.stop();
    		finished = true;
    	} else {
    		if(Math.abs(angleDifference) < 10) {
    			turnRate = 0.5;
    		} else { 
    			turnRate = 0.3;
    		}
    	}
    	
    	if(angleDifference < 0) {
    		turnRate = -turnRate;
    	}
    	
    	Robot.chassis.arcadeDrive(0.0, turnRate);
    	
    }

    // Make this return true when this Command no longer needs to run execute()
    protected boolean isFinished() {
        return finished;
    }

    // Called once after isFinished returns true
    protected void end() {
    	System.out.println("[TurnSpecifiedDegreesCommand] completed.");
    	Robot.chassis.stop();
    }

    // Called when another command which requires one or more of the same
    // subsystems is scheduled to run
    protected void interrupted() {
    	System.out.println("[TurnSpecifiedDegreesCommand] Interrupted.");
    	Robot.chassis.stop();
    }
}
